asterisk_click2dial/models/asterisk_server.py

```diff
@@ -0,0 +1,283 @@
+                (self.ip_address, self.port), self.login, self.password
+            )
+        except Exception as e:
+            raise UserError(_("Connection Test Failed! The error message is: %s" % e))
+        finally:
+            if ast_manager:
+                ast_manager.Logoff()
+        raise UserError(
+            _(
+                "Connection Test Successfull! Odoo can successfully login to "
+                "the Asterisk Manager Interface."
+            )
+        )
+
+    @api.model
+    def _get_calling_number_from_channel(self, chan, user):
+        """Method designed to be inherited to work with
+        very old or very new versions of Asterisk"""
+        sip_account = user.asterisk_chan_type + "/" + user.resource
+        internal_number = user.internal_number
+        # 4 = Ring
+        # 6 = Up
+        if chan.get("ChannelState") in ("4", "6") and (
+            chan.get("ConnectedLineNum") == internal_number
+            or chan.get("EffectiveConnectedLineNum") == internal_number
+            or sip_account in chan.get("BridgedChannel", "")
+        ):
+            _logger.debug(
+                "Found a matching Event with channelstate = %s",
+                chan.get("ChannelState"),
+            )
+            return chan.get("CallerIDNum")
+        # Compatibility with Asterisk 1.4
+        if chan.get("State") == "Up" and sip_account in chan.get("Link", ""):
+            _logger.debug("Found a matching Event in 'Up' state")
+            return chan.get("CallerIDNum")
+        return False
+
+    @api.model
+    def _get_calling_number(self):
+        user, ast_server, ast_manager = self._connect_to_asterisk()
+        calling_party_number = False
+        try:
+            list_chan = ast_manager.Status()
+            # from pprint import pprint
+            # pprint(list_chan)
+            _logger.debug("Result of Status AMI request:")
+            _logger.debug(pformat(list_chan))
+            for chan in list_chan.values():
+                calling_party_number = self._get_calling_number_from_channel(chan, user)
+                if calling_party_number:
+                    break
+        except Exception as e:
+            _logger.error(
+                "Error in the Status request to Asterisk server %s",
+                ast_server.ip_address,
+            )
+            _logger.error("Here are the details of the error: '%s'", str(e))
+            raise UserError(
+                _(
+                    "Can't get calling number from  Asterisk.\nHere is the "
+                    "error: '%s'" % str(e)
+                )
+            )
+
+        finally:
+            ast_manager.Logoff()
+
+        _logger.debug("Calling party number: '%s'", calling_party_number)
+        return calling_party_number
+
+    @api.model
+    def get_record_from_my_channel(self):
+        calling_number = self.env["asterisk.server"]._get_calling_number()
+        # calling_number = "0641981246"
```

Comment:
Is that a valid number for testing? Maybe remove it :)

Reply:
ok. It just help me locate the place to put my test number.